feat(tabs): adds the `md-tab-group` component

robertmesserle opened the pull request:
@jelbourn @kara @hansl

kara commented on src/components/tab-group/tab-group.spec.ts:
```diff
@@ -0,0 +1,123 @@
+import {By} from 'angular2/platform/browser';
+
+export function main() {
+  ddescribe('MdTabGroup', () => {
```
Looks like you forgot a `ddescribe` here :)

kara commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,65 @@
+   * @param i
+   * @returns {string}
+   */
+  getTabLabelId(i: number) {
```
Put the return type on the function itself?

kara commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,65 @@
+   * @returns {string}
+   */
+  getTabLabelId(i: number) {
+    return 'md-tab-label-' + this._groupId + '-' + i;
```
It seems like this would be more readable with a template string, what do you think?

```typescript
return `md-tab-label-${this._groupId}-${i}`;
```

kara commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,65 @@
+  /**
+   * Increment the selected index by 1; prevent going over the number of tabs
+   */
+  nextTab() {
```
These `nextTab` and `previousTab` functions look similar to the methods in `TabBar`.  Any way we can re-use?

robertmesserle replied:
These will most-likely be reused when tab-bar is implemented, but this isn't in this iteration.  I only left this file in because Jeremy said to leave it in.

kara commented on src/components/tab-group/tab-group.scss:
```diff
@@ -0,0 +1,61 @@
+
+:host {
+  display: block;
+  font-family: sans-serif;
```
Seems like maybe this should use the common `$font-family` variable? Or should this font be different?

kara commented on src/components/tab-group/tab-group.scss:
```diff
@@ -0,0 +1,61 @@
+.md-tab-label {
+  float: left;
+  line-height: 48px;
+  height: 48px;
```
Have you considered putting this into a Sass variable?  Might help make it more customizable.

robertmesserle commented:
@kara Addressed your comments.

@jelbourn This might actually be okay to merge this week - I got it working with focus state and modified the arrow functionality so that it cycles focus target.  Enter will actually select the tab.

princemaple commented on src/components/tab-group/tab-group.html:
```diff
@@ -0,0 +1,25 @@
+     (keydown.arrowLeft)="focusPreviousTab()"
+     (keydown.enter)="selectedIndex = focusIndex">
+  <div class="md-tab-label" role="tab" md-tab-label-wrapper
+       *ngFor="#label of labels; #i = index"
```
let

princemaple commented on src/components/tab-group/tab-group.html:
```diff
@@ -0,0 +1,25 @@
+</div>
+<div class="md-tab-body-wrapper">
+  <div class="md-tab-body"
+       *ngFor="#content of contents; #i = index"
```
let - ng2 beta 17

jelbourn commented:
@robertmesserle Now needs rebasing on my mega change #384

jelbourn commented on src/components/tab-group/tab-group.html:
```diff
@@ -0,0 +1,25 @@
+  <div class="md-tab-body"
+       *ngFor="let content of contents; let i = index"
+       [class.md-active]="selectedIndex == i"
+       [attr.id]="getTabContentId(i)"
```
Just `id` should work.

jelbourn commented on src/components/tab-group/ink-bar.ts:
```diff
@@ -0,0 +1,23 @@
+import {Directive, Renderer, ElementRef} from '@angular/core';
+
+@Directive({
```
Add class description explaining what the ink bar is and how it's used.

jelbourn replied:
You probably also want to add an `@internal` JsDoc; this makes TypeScript ignore it when generating the `.d.ts.`. For example:

```ts
/**
 * This is the description of the ...
 * @internal
 */
```

jelbourn commented on src/components/tab-group/ink-bar.ts:
```diff
@@ -0,0 +1,23 @@
+  selector: 'md-ink-bar',
+})
+export class MdInkBar {
+  constructor(public renderer: Renderer, public elementRef: ElementRef) {}
```
These fields should probably be `private`

jelbourn commented on src/components/tab-group/ink-bar.ts:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  getLeftPosition(element: HTMLElement): string {
+    return element.offsetLeft + 'px';
```
Does `ElementRef` actually have `offsetLeft` and `offsetWidth` properties? I think you actually need to pull them through the renderer.

jelbourn commented on src/components/tab-group/tab-content.ts:
```diff
@@ -0,0 +1,11 @@
+import {Directive, TemplateRef, ViewContainerRef} from '@angular/core';
+import {TemplatePortalDirective} from '../../core/portal/portal-directives';
+
+@Directive({
```
Add class description (here and for other classes throughout the PR)

jelbourn commented on src/components/tab-group/tab-group.html:
```diff
@@ -0,0 +1,25 @@
+       [id]="getTabLabelId(i)"
+       [attr.aria-controls]="getTabContentId(i)"
+       [attr.aria-selected]="selectedIndex == i"
+       [attr.tabindex]="selectedIndex == i ? 0 : -1"
```
You should be able to bind `tabIndex` without `attr.`

jelbourn commented on src/components/tab-group/tab-group.scss:
```diff
@@ -0,0 +1,66 @@
+@import 'variables';
+@import 'default-theme';
+
+$tab-bar-height: 48px !default;
```
Prefix scss vars with `md-`

jelbourn commented on src/components/tab-group/tab-group.scss:
```diff
@@ -0,0 +1,66 @@
+  }
+}
+
+.md-tab-header {
```
For each of the css classes here, could you add a brief overview of what they're for or which element/state they correspond to?

jelbourn commented on src/components/tab-group/tab-group.scss:
```diff
@@ -0,0 +1,66 @@
+  position: relative;
+  display: flex;
+  flex-direction: row;
+  &:before, &:after {
```
Add comment explaining what the `:before` and `:after` are doing here?

jelbourn commented on src/components/tab-group/tab-group.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  `,
+  directives: [MD_TAB_GROUP_DIRECTIVES]
+})
+class TestApp {
```
I'd be more specific with this name, like `TabsTestApp`

jelbourn commented on src/components/tab-group/tab-group.spec.ts:
```diff
@@ -0,0 +1,134 @@
+      checkSelectedIndex(fixture, 0);
+
+      // select the second tab
+      let tabLabel = fixture.debugElement.query(By.css('.md-tab-label:nth-of-type(2)'));
```
It will probably be much cleaner to do the `createAsync` in a beforeEach, and then capture all of the tab labels in an array and access them directly in each test. I'm doing in in my in-progress rewrite of the radio button tests, I can show you if you're curious.

jelbourn commented on src/components/tab-group/tab-group.spec.ts:
```diff
@@ -0,0 +1,134 @@
+    builder = tcb;
+  }));
+
+  it('should default to the first tab', (done: () => void) => {
```
Rather than passing in `done` and calling it manually, you can now use `async` when dealing with `createAsync`, e.g.:
```ts
it('should do something', async(() => {
  builder.createAsync(TestApp).then(fixture => {
    expect(1).toBe(1);
  });
}));
```
`async()` hooks into zones and makes it so that the test doesn't "complete" until all of the microtasks have been executed.

jelbourn commented on src/components/tab-group/tab-group.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  });
+});
+
+function checkSelectedIndex(fixture: ComponentFixture<TestApp>, index: number) {
```
Add function description

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+
+/**
+ * Used to generate unique ID's for each tab component
+ * @type {number}
```
Don't need `@type`, it's in the language now

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+  @ContentChildren(MdTabLabel) labels: QueryList<MdTabLabel>;
+  @ContentChildren(MdTabContent) contents: QueryList<MdTabContent>;
+  @ViewChildren(MdTabLabelWrapper) labelWrappers: QueryList<MdTabLabelWrapper>;
+  @ViewChildren(MdInkBar) inkBar: QueryList<MdInkBar>;
```
Each of these `__Children` properties should probably be `private`

jelbourn replied:
You can use `@ViewChild` instead of `@ViewChildren` to just get a single instance.
```ts
 @ViewChild(MdInkBar) inkBar: MdInkBar;
```

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+   * Note: This must be run outside of the zone or it will create an infinite change detection loop
+   */
+  ngAfterViewChecked(): void {
+    this._zone.runOutsideAngular(() => {
```
Add explanation of why you're doing the `runOutsideAngular`

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+   */
+  ngAfterViewChecked(): void {
+    this._zone.runOutsideAngular(() => {
+      window.requestAnimationFrame(() => {
```
It should work to just say `requestAnimationFrame` without explicitly dereferencing `window`.

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+   * ViewChildren references are ready.
+   * @returns {HTMLElement | null}
+   */
+  get currentLabelWrapper(): HTMLElement {
```
I think this should just be a function rather than a getter property.

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+
+  /**
+   * Tracks which element has focus; used for keyboard navigation
+   * @returns {number}
```
You don't need the type in the JsDoc

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  /**
+   * Tracks which element has focus; used for keyboard navigation
```
```ts
/** Gets the index of the tab that currently has focus. */
```
?

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+   * @param i
+   * @returns {string}
+   */
+  getTabLabelId(i: number): string {
```
Could the labels and contents generate their own IDs?

jelbourn commented on src/components/tab-group/tab-group.ts:
```diff
@@ -0,0 +1,120 @@
+   * Decrement the focus index by 1; prevent going below 0
+   */
+  focusPreviousTab(): void {
+    if (this.focusIndex > 0) {
```
optional: could do
```ts
this.focusIndex = Math.max(0, this.focusIndex - 1);
```

jelbourn commented on src/components/tab-group/tab-group.scss:
```diff
@@ -0,0 +1,66 @@
+  display: block;
+  font-family: $md-font-family;
+  .md-tab-header {
+    border-bottom: 1px solid md-color($md-background, status-bar);
```
Why is this one rule for `md-tab-header` here instead of in the rule below?

jelbourn commented:
Looks good aside from that one last comment and a needed rebase.

jelbourn commented:
LGTM
